compiler/rustc_data_structures/src/sync.rs

```diff
@@ -299,48 +340,125 @@ cfg_if! {
         use parking_lot::RwLock as InnerRwLock;
 
         use std::thread;
-        pub use rayon::{join, scope};
+
+        #[inline]
+        pub fn join<A, B, RA: Send, RB: Send>(oper_a: A, oper_b: B) -> (RA, RB)
+        where
+            A: FnOnce() -> RA + DynSend + DynSync,
+            B: FnOnce() -> RB + DynSend + DynSync,
+        {
+            if is_parallel() {
+                let oper_a = IsSend::new(oper_a);
+                let oper_b = IsSend::new(oper_b);
+                rayon::join(move || oper_a.into_inner()(), move || oper_b.into_inner()())
+            } else {
+                (oper_a(), oper_b())
+            }
+        }
+
+        pub use rayon::scope;
 
         /// Runs a list of blocks in parallel. The first block is executed immediately on
         /// the current thread. Use that for the longest running block.
         #[macro_export]
         macro_rules! parallel {
-            (impl $fblock:tt [$($c:tt,)*] [$block:tt $(, $rest:tt)*]) => {
-                parallel!(impl $fblock [$block, $($c,)*] [$($rest),*])
+            ([$dyn_send: expr], impl $fblock:tt [$($c:tt,)*] [$block:tt $(, $rest:tt)*]) => {
+                parallel!([$dyn_send], impl $fblock [$block, $($c,)*] [$($rest),*])
+            };
+            ([$dyn_send: expr], impl $fblock:tt [$($blocks:tt,)*] []) => {
+                {
+                    let dyn_send = IsSend::new($dyn_send);
+                    ::rustc_data_structures::sync::scope(|s| {
+                        $(
+                            s.spawn(|_| $blocks(*dyn_send));
+                        )*
+                        $fblock(*dyn_send);
+                    })
+                }
             };
-            (impl $fblock:tt [$($blocks:tt,)*] []) => {
-                ::rustc_data_structures::sync::scope(|s| {
+            ([$dyn_send: expr], $fblock:expr, $($blocks:expr),*) => {
+                if is_parallel() {
+                    // Reverse the order of the later blocks since Rayon executes them in reverse order
+                    // when using a single thread. This ensures the execution order matches that
+                    // of a single threaded rustc
+                    parallel!([$dyn_send], impl $fblock [] [$($blocks),*]);
+                } else {
+                    // We catch panics here ensuring that all the blocks execute.
+                    // This makes behavior consistent with the parallel compiler.
+                    let mut ps: Vec<_> = Vec::new();
                     $(
-                        s.spawn(|_| $blocks);
```

Comment:
You should wrap this closure in `IsSend` instead of adding the `$dyn_send` argument to this macro.

Reply:
Thanks! It's much cleaner now!